### PR TITLE
Optimize block decoding

### PR DESCRIPTION
**Pull request by tdcmeehan**
Use memory copies when nulls aren't present

### CODE REVIEW

**Comment by rongrong** [comment]
Would be interesting to see benchmark numbers. ^_^

**Comment by tdcmeehan** [comment]
Added a quick benchmark for `BIGINT`.

Before:

```
Benchmark                                      (channelCount)  (hasNull)  (type)  Mode  Cnt   Score   Error  Units
BenchmarkLongArrayBlockEncoding.testReadBlock               1       true  BIGINT  avgt   30  47.135 ± 1.465  us/op
BenchmarkLongArrayBlockEncoding.testReadBlock               1      false  BIGINT  avgt   30  38.394 ± 0.834  us/op
BenchmarkLongArrayBlockEncoding.testReadBlock               2       true  BIGINT  avgt   30  72.875 ± 1.708  us/op
BenchmarkLongArrayBlockEncoding.testReadBlock               2      false  BIGINT  avgt   30  57.430 ± 1.962  us/op
```

After:

```
Benchmark                                      (channelCount)  (hasNull)  (type)  Mode  Cnt   Score   Error  Units
BenchmarkLongArrayBlockEncoding.testReadBlock               1       true  BIGINT  avgt   30  30.400 ± 1.440  us/op
BenchmarkLongArrayBlockEncoding.testReadBlock               1      false  BIGINT  avgt   30  12.430 ± 0.742  us/op
BenchmarkLongArrayBlockEncoding.testReadBlock               2       true  BIGINT  avgt   30  50.845 ± 2.144  us/op
BenchmarkLongArrayBlockEncoding.testReadBlock               2      false  BIGINT  avgt   30  19.241 ± 1.037  us/op
```

So on the microbenchmark we see a 30% improvement with nulls, 66% improvement without nulls.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/EncoderUtil.java [hunk 1]
```diff
@@ -97,4 +100,41 @@ public static void encodeNullsAsBits(SliceOutput sliceOutput, Block block)
 
         return Optional.of(valueIsNull);
     }
+
+    /**
+     * Decode the bit stream created by encodeNullsAsBits.  The valueIsNull buffer must be exactly sized to the position
+     * count.  Returns the number of nulls seen during deserialization.
+     */
+    public static int decodeNullBits(SliceInput sliceInput, boolean[] valueIsNull)
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/EncoderUtil.java]: Do we want to make the `decodeNullBits` method at line 71 to call this method?

Comment (tdcmeehan) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/EncoderUtil.java]: There's a very slight performance regression in decoding the null bits for array and map blocks if we combine them...

Comment (wenleix) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/EncoderUtil.java]: @tdcmeehan : hmm.. why ? These two methods are slightly different? Then maybe adding comment ? ;)

**Comment by tdcmeehan** [comment]
Once #14329 this should improve performance on the raw page format further, will benchmark.

**Comment by tdcmeehan** [comment]
Performance results for HiveFileFormatBenchmark on `PRESTO_PAGE`:  https://gist.github.com/tdcmeehan/a102aaf5cb8be789d80c62b6df17e6c5

Tl;dr, most dramatic improvement is seen for primitives, maps of primitives also see speedup, no significant movement on variable width types (as this is not optimized).

**Comment by tdcmeehan** [comment]
Optimizations for writes can go on a different PR.

File: presto-common/src/main/java/com/facebook/presto/common/block/ByteArrayBlockEncoding.java [hunk 2]
```diff
@@ -49,16 +51,27 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
...
-
         byte[] values = new byte[positionCount];
-        for (int position = 0; position < positionCount; position++) {
-            if (valueIsNull == null || !valueIsNull[position]) {
-                values[position] = sliceInput.readByte();
-            }
+
+        // Fast track if no nulls present
+        if (!sliceInput.readBoolean()) {
+            sliceInput.readBytes(wrappedBuffer(values));
```
Comment (yingsu00) [inline presto-common/src/main/java/com/facebook/presto/common/block/ByteArrayBlockEncoding.java:58]: There is no need to create a new Slice object by wrappedBuffer. There is `public final void readBytes(byte[] destination)` in SliceInput.

File: presto-common/src/main/java/com/facebook/presto/common/block/ByteArrayBlockEncoding.java [hunk 3]
```diff
@@ -49,16 +51,27 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
...
+            return new ByteArrayBlock(0, positionCount, null, values);
         }
 
+        boolean[] valueIsNull = new boolean[positionCount];
+        int nullPositions = decodeNullBits(sliceInput, valueIsNull);
+        int nonNullPositions = positionCount - nullPositions;
+
+        // Read compact values array and redistribute to non-null positions
+        sliceInput.readBytes(wrappedBuffer(values), nonNullPositions * SIZE_OF_BYTE);
```
Comment (yingsu00) [inline presto-common/src/main/java/com/facebook/presto/common/block/ByteArrayBlockEncoding.java:67]: Use `public abstract void readBytes(byte[] destination, int destinationIndex, int length);`. Creating many Slice objects could hurt performance and reliability.

File: presto-common/src/main/java/com/facebook/presto/common/block/EncoderUtil.java [hunk 4]
```diff
@@ -97,4 +100,41 @@ public static void encodeNullsAsBits(SliceOutput sliceOutput, Block block)
 
         return Optional.of(valueIsNull);
     }
+
+    /**
+     * Decode the bit stream created by encodeNullsAsBits.  The valueIsNull buffer must be exactly sized to the position
+     * count.  Returns the number of nulls seen during deserialization.
+     */
+    public static int decodeNullBits(SliceInput sliceInput, boolean[] valueIsNull)
```
Comment (yingsu00) [inline presto-common/src/main/java/com/facebook/presto/common/block/EncoderUtil.java:108]: This new method shares 90% same code as existing `public static Optional<boolean[]> decodeNullBits(SliceInput sliceInput, int positionCount)`. Can we extract the common part into a method to reduce code duplication?

Comment (yingsu00) [reply presto-common/src/main/java/com/facebook/presto/common/block/EncoderUtil.java:108]: Emm, I see the new method is doing bitCount on the fly. So it makes sense. Please disregard this comment.

**Comment by tdcmeehan** [comment]
@yingsu00 how about we add `readInts`\`readLongs`\etc method to `SliceInput`?  That way we could avoid the intermediate slice creation and directly use unsafe to perform a memory copy.

**Comment by yingsu00** [comment]
> @yingsu00 how about we add `readInts` \ `readLongs` \ etc method to `SliceInput`? That way we could avoid the intermediate slice creation and directly use unsafe to perform a memory copy.

That'll be nice. But we need to fork Slice first :)

File: presto-common/src/main/java/com/facebook/presto/common/block/ByteArrayBlockEncoding.java [hunk 5]
```diff
@@ -49,16 +50,27 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
...
+        if (!sliceInput.readBoolean()) {
+            sliceInput.readBytes(values);
+            return new ByteArrayBlock(0, positionCount, null, values);
         }
 
+        boolean[] valueIsNull = new boolean[positionCount];
+        int nullPositions = decodeNullBits(sliceInput, valueIsNull);
+        int nonNullPositions = positionCount - nullPositions;
+
+        // Read compact values array and redistribute to non-null positions
```
Comment (mbasmanova) [inline presto-common/src/main/java/com/facebook/presto/common/block/ByteArrayBlockEncoding.java:65]: This looks very similar to com.facebook.presto.orc.reader.ReaderUtils#unpackByteNulls(byte[], boolean[])

Consider refactoring to avoid copy-paste.

File: presto-common/src/main/java/com/facebook/presto/common/block/EncoderUtil.java (hunk 4 above)
Comment (mbasmanova) [inline presto-common/src/main/java/com/facebook/presto/common/block/EncoderUtil.java:108]: Consider refactoring to avoid so much copy-paste between this new version and existing one.
